File: runtime-app/lib/models/tenant_config.dart
class BuildManifest {
  const BuildManifest({
    required this.capabilityKeys,
    required this.flutterPkgs,
    required this.webPkgs,
    required this.routes,
    required this.deliver,
  });

  final List<String> capabilityKeys;
  final List<String> flutterPkgs;
  final List<String> webPkgs;
  final List<String> routes;
  final String deliver;

  factory BuildManifest.fromJson(Map<String, dynamic> json) => BuildManifest(
        capabilityKeys: (json['capability_keys'] as List<dynamic>? ?? [])
            .map((e) => e.toString())
            .toList(),
        flutterPkgs: (json['flutter_pkgs'] as List<dynamic>? ?? [])
            .map((e) => e.toString())
            .toList(),
        webPkgs: (json['web_pkgs'] as List<dynamic>? ?? [])
            .map((e) => e.toString())
            .toList(),
        routes: (json['routes'] as List<dynamic>? ?? [])
            .map((e) => e.toString())
            .toList(),
        deliver: json['deliver'] as String? ?? 'both',
      );
}

class PageSchema {
  const PageSchema({
    required this.version,
    required this.menu,
    required this.capabilityKeys,
    this.root,
  });

  final String version;
  final List<MenuItem> menu;
  final List<String> capabilityKeys;
  final Map<String, dynamic>? root;

  factory PageSchema.fromJson(Map<String, dynamic> json) {
    final menuRaw = json['menu'] as List<dynamic>? ?? [];
    final keysRaw = json['capability_keys'] as List<dynamic>? ?? [];
    return PageSchema(
      version: json['version'] as String? ?? '1',
      menu: menuRaw
          .map((e) => MenuItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      capabilityKeys: keysRaw.map((e) => e.toString()).toList(),
      root: json['root'] as Map<String, dynamic>?,
    );
  }
}

class TenantConfig {
  const TenantConfig({
    required this.tenantSlug,
    required this.tenantName,
    required this.appName,
    required this.primaryColor,
    required this.theme,
    required this.apiBaseUrl,
    required this.menu,
    this.appIconUrl,
    this.features = const {},
    this.app,
    this.pageSchema,
    this.buildManifest,
  });

  final String tenantSlug;
  final String tenantName;
  final String appName;
  final String? appIconUrl;
  final String primaryColor;
  final String theme;
  final String apiBaseUrl;
  final List<MenuItem> menu;
  final Map<String, bool> features;
  final AppInfo? app;
  final PageSchema? pageSchema;
  final BuildManifest? buildManifest;

  /// 契约驱动的 capability_key 列表（优先 manifest，其次 page_schema，最后 app）。
  List<String> get resolvedCapabilityKeys {
    if (buildManifest != null && buildManifest!.capabilityKeys.isNotEmpty) {
      return buildManifest!.capabilityKeys;
    }
    if (pageSchema != null && pageSchema!.capabilityKeys.isNotEmpty) {
      return pageSchema!.capabilityKeys;
    }
    return app?.capabilityKeys ?? [];
  }

  factory TenantConfig.fromJson(Map<String, dynamic> json) {
    final menuRaw = json['menu'] as List<dynamic>? ?? [];
    PageSchema? pageSchema;
    BuildManifest? buildManifest;

    final appJson = json['app'] as Map<String, dynamic>?;
    if (appJson != null) {
      final schemaRaw = appJson['page_schema'] as Map<String, dynamic>?;
      if (schemaRaw != null) {
        pageSchema = PageSchema.fromJson(schemaRaw);
      }
      final manifestRaw = appJson['build_manifest'] as Map<String, dynamic>?;
      if (manifestRaw != null) {
        buildManifest = BuildManifest.fromJson(manifestRaw);
      }
    }

    return TenantConfig(
      tenantSlug: json['tenant_slug'] as String? ?? 'demo',
      tenantName: json['tenant_name'] as String? ?? '',
      appName: json['app_name'] as String? ?? 'TrackChat',
      appIconUrl: json['app_icon_url'] as String?,
      primaryColor: json['primary_color'] as String? ?? '#4338CA',
      theme: json['theme'] as String? ?? 'light',
      apiBaseUrl: json['api_base_url'] as String? ?? '',
      menu: menuRaw
          .map((e) => MenuItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      features: (json['features'] as Map<String, dynamic>? ?? {})
          .map((k, v) => MapEntry(k, v == true)),
      app: appJson != null ? AppInfo.fromJson(appJson) : null,
      pageSchema: pageSchema,
      buildManifest: buildManifest,
    );
  }
}

class MenuItem {
  const MenuItem({required this.key, required this.label, required this.icon});

  final String key;
  final String label;
  final String icon;

  factory MenuItem.fromJson(Map<String, dynamic> json) => MenuItem(
        key: json['key'] as String? ?? '',
        label: json['label'] as String? ?? '',
        icon: json['icon'] as String? ?? '',
      );
}

class AppInfo {
  const AppInfo({
    required this.id,
    required this.name,
    required this.schemaUrl,
    required this.modules,
    required this.capabilityKeys,
  });

  final String id;
  final String name;
  final String schemaUrl;
  final List<dynamic> modules;
  final List<String> capabilityKeys;

  factory AppInfo.fromJson(Map<String, dynamic> json) => AppInfo(
        id: json['id'] as String? ?? '',
        name: json['name'] as String? ?? '',
        schemaUrl: json['schema_url'] as String? ?? '',
        modules: json['modules'] as List<dynamic>? ?? [],
        capabilityKeys: (json['capability_keys'] as List<dynamic>? ?? [])
            .map((e) => e.toString())
            .toList(),
      );
}

File: backend/app/services/tenant_config.py
"""Tenant runtime config for Web / Flutter clients."""

from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from app.core.config import settings
from app.db.models import AppRecord, ApprovalRecord, Tenant, User
from app.services.apk_builder import get_apk_build_status, per_app_apk_ready
from app.services.db_seed import DEFAULT_TENANT_SLUG

DEFAULT_TENANT_CONFIG: dict[str, Any] = {
    "app_name": "TrackChat",
    "app_icon_url": "",
    "primary_color": "#4338ca",
    "theme": "light",
    "api_base_url": "",
    "menu": [
        {"key": "home", "label": "首页", "icon": "home"},
        {"key": "chat", "label": "智能问答", "icon": "chat"},
        {"key": "approval", "label": "审批", "icon": "approval"},
    ],
    "features": {
        "chat_sse": True,
        "offline_cache": False,
    },
}


def _merge_config(base: dict[str, Any], override: dict[str, Any] | None) -> dict[str, Any]:
    if not override:
        return dict(base)
    merged = dict(base)
    for key, value in override.items():
        if isinstance(value, dict) and isinstance(merged.get(key), dict):
            merged[key] = {**merged[key], **value}
        else:
            merged[key] = value
    return merged


def ensure_tenant_config(db: Session, tenant: Tenant) -> dict[str, Any]:
    base = dict(DEFAULT_TENANT_CONFIG)
    base["api_base_url"] = f"{settings.public_base_url.rstrip('/')}{settings.api_prefix}"
    if tenant.config_json:
        return _merge_config(base, tenant.config_json)
    tenant.config_json = base
    db.add(tenant)
    db.commit()
    db.refresh(tenant)
    return tenant.config_json or base


def get_tenant_config(
    db: Session,
    *,
    tenant_slug: str = DEFAULT_TENANT_SLUG,
    app_public_id: str | None = None,
) -> dict[str, Any]:
    tenant = db.query(Tenant).filter(Tenant.slug == tenant_slug).first()
    if not tenant:
        tenant = Tenant(name="TrackChat 演示租户", slug=tenant_slug)
        db.add(tenant)
        db.flush()

    config = ensure_tenant_config(db, tenant)
    payload: dict[str, Any] = {
        "tenant_slug": tenant.slug,
        "tenant_name": tenant.name,
        **config,
    }

    # 按应用打开运行时时跳过全局 COUNT（三表全表统计会拖慢 /config）
    if not app_public_id:
        apps_count = db.query(AppRecord).filter(AppRecord.tenant_id == tenant.id).count()
        approvals_count = db.query(ApprovalRecord).filter(ApprovalRecord.tenant_id == tenant.id).count()
        users_count = db.query(User).filter(User.tenant_id == tenant.id).count()
        payload["stats"] = {
            "apps": apps_count,
            "approvals": approvals_count,
            "users": users_count,
        }

    if app_public_id:
        app = (
            db.query(AppRecord)
            .filter(AppRecord.public_id == app_public_id, AppRecord.tenant_id == tenant.id)
            .first()
        )
        if app:
            deliver = app.deliver or "both"
            base_url = settings.public_base_url.rstrip("/")
            # 运行时壳只需要文件是否存在；不读 build log / detail
            apk_ready = per_app_apk_ready(app.public_id, deliver=deliver)
            apk_status = get_apk_build_status(app.public_id) if deliver in ("app", "both") else "skipped"
            payload["app"] = {
                "id": app.public_id,
                "name": app.name,
                "schema_url": app.schema_url,
                "deliver": deliver,
                "modules": app.modules,
                "capability_keys": app.capability_keys,
                "web_url": f"{base_url}/r/{app.public_id}",
                "download_url": f"{base_url}/r/{app.public_id}/download",
                "apk_ready": apk_ready,
                "apk_build_status": apk_status,
            }
            payload["deliver"] = deliver
            payload["web_url"] = f"{base_url}/r/{app.public_id}"
            payload["download_url"] = f"{base_url}/r/{app.public_id}/download"
            payload["apk_ready"] = apk_ready
            payload["apk_build_status"] = apk_status
            payload["app_name"] = app.name
            if app.icon_url:
                payload["app_icon_url"] = app.icon_url
            if app.primary_color:
                payload["primary_color"] = app.primary_color
            if app.page_schema and isinstance(app.page_schema, dict):
                menu = app.page_schema.get("menu")
                if menu:
                    payload["menu"] = menu

    return payload


def update_tenant_config(
    db: Session,
    *,
    tenant_slug: str,
    patch: dict[str, Any],
) -> dict[str, Any]:
    tenant = db.query(Tenant).filter(Tenant.slug == tenant_slug).first()
    if not tenant:
        tenant = Tenant(name="TrackChat 演示租户", slug=tenant_slug)
        db.add(tenant)
        db.flush()

    allowed_keys = {
        "app_name",
        "app_icon_url",
        "primary_color",
        "theme",
        "menu",
        "features",
    }
    current = ensure_tenant_config(db, tenant)
    merged = dict(current)
    for key, value in patch.items():
        if key not in allowed_keys:
            continue
        if key == "menu" and isinstance(value, list):
            merged["menu"] = value
        elif key == "features" and isinstance(value, dict):
            merged["features"] = {**merged.get("features", {}), **value}
        else:
            merged[key] = value

    tenant.config_json = merged
    db.add(tenant)
    db.commit()
    db.refresh(tenant)
    return get_tenant_config(db, tenant_slug=tenant_slug)
